### PR TITLE
Use `goverter` to automatically generate conversion code

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane/crossplane/issues/3274
Fixes https://github.com/crossplane/crossplane/issues/2499

Using [`goverter`](https://github.com/jmattheis/goverter) for this purpose is particularly nice because its default behaviour is to fail with an error if asked to generate code that converts from one type to another that is not identical. This means that for example generating the `RevisionSpecConverter` will fail if someone adds a new field to `*v1.CompositionSpec` without also adding it to `*v1.beta1.CompositionRevisionSpec` (or vice versa).

The PR also contains some related commits:

* I've added a shell script that "generates" identical API types. This is useful when we promote a type from (e.g.) v1beta1 to v1; we can ensure folks don't forget to update the "old" types because they'll be dumb copies of the newer types.
* I've deleted the `Lock` type from `pkg.crossplane.io/v1alpha1`. I was going to use said script to copy it automatically, but noticed we'd marked it as due for deletion in v1.7 (several releases ago). I've added the `breaking-api-change` label to draw attention to this, despite removing an alpha API technically not being a "breaking change".

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

[contribution process]: https://git.io/

I confirmed that the existing conversion tests passed before I deleted them (to avoid testing generated code). I've also tested this using https://github.com/upbound/platform-ref-gcp. This doesn't exercise every single Composition feature, but I did pay close attention to the CompositionRevisions that were generated to ensure it appeared the generated conversion code was working. As far as I can tell the revisions are as expected - all the patches, transforms, etc were propagated from composition to revision.

### CODE REVIEW

**Comment by negz** [comment]
Here's what happens if you try to add a new type to `CompositionSpec` without also adding it to `CompositionRevisionSpec`. First, the relevant diff:

```diff
--- a/apis/apiextensions/v1/composition_transforms.go
+++ b/apis/apiextensions/v1/composition_transforms.go
@@ -344,6 +344,7 @@ const (
        StringConversionTypeToLower    StringConversionType = "ToLower"
        StringConversionTypeToBase64   StringConversionType = "ToBase64"
        StringConversionTypeFromBase64 StringConversionType = "FromBase64"
+       StringConversionTypeBreak      StringConversionType = "BreakGoverter"
 )

 // A StringTransform returns a string given the supplied input.
@@ -351,7 +352,7 @@ type StringTransform struct {

        // Type of the string transform to be run.
        // +optional
-       // +kubebuilder:validation:Enum=Format;Convert;TrimPrefix;TrimSuffix;Regexp
+       // +kubebuilder:validation:Enum=Format;Convert;TrimPrefix;TrimSuffix;Regexp;BreakGoverter
        // +kubebuilder:default=Format
        Type StringTransformType `json:"type,omitempty"`

@@ -372,6 +373,16 @@ type StringTransform struct {
        // Extract a match from the input using a regular expression.
        // +optional
        Regexp *StringTransformRegexp `json:"regexp,omitempty"`
+
+       // Break goverter!
+       // +optional
+       Break *StringTransformBreak `json:"break,omitempty"`
+}
+
+// StringTransformBreak breaks goverter!
+type StringTransformBreak struct {
+       // BreakIt specifies whether we should break it.
+       BreakIt bool `json:"breakIt"`
 }

 // A StringTransformRegexp extracts a match from the input using a regular
```

This results in the following behaviour at build time:

```console
$ make reviewable
21:14:11 [ .. ] verify go modules dependencies have expected content
all modules verified
21:14:13 [ OK ] go modules dependencies verified
21:14:13 [ .. ] go generate linux_arm64
Duplicated apiextensions/v1beta1/revision_types.go (package v1beta1) to apiextensions/v1alpha1/zz_generated.revision_types.go (package v1alpha1).
Duplicated pkg/meta/v1/configuration_types.go (package v1) to pkg/meta/v1alpha1/zz_generated.configuration_types.go (package v1alpha1).
Duplicated pkg/meta/v1/provider_types.go (package v1) to pkg/meta/v1alpha1/zz_generated.provider_types.go (package v1alpha1).
Duplicated pkg/meta/v1/meta.go (package v1) to pkg/meta/v1alpha1/zz_generated.meta.go (package v1alpha1).
Error while creating converter method:
    func (github.com/crossplane/crossplane/apis/apiextensions/v1.RevisionSpecConverter).FromRevisionSpec(in github.com/crossplane/crossplane/apis/apiextensions/v1beta1.CompositionRevisionSpec) github.com/crossplane/crossplane/apis/api
extensions/v1.CompositionSpec

| github.com/crossplane/crossplane/apis/apiextensions/v1beta1.CompositionRevisionSpec
|
|      | []github.com/crossplane/crossplane/apis/apiextensions/v1beta1.PatchSet
|      |
|      |        | github.com/crossplane/crossplane/apis/apiextensions/v1beta1.PatchSet
|      |        |
|      |        |  | []github.com/crossplane/crossplane/apis/apiextensions/v1beta1.Patch
|      |        |  |
|      |        |  |      | github.com/crossplane/crossplane/apis/apiextensions/v1beta1.Patch
|      |        |  |      |
|      |        |  |      |  | []github.com/crossplane/crossplane/apis/apiextensions/v1beta1.Transform
|      |        |  |      |  |
|      |        |  |      |  |         | github.com/crossplane/crossplane/apis/apiextensions/v1beta1.Transform
|      |        |  |      |  |         |
|      |        |  |      |  |         |  | *github.com/crossplane/crossplane/apis/apiextensions/v1beta1.StringTransform
|      |        |  |      |  |         |  |
|      |        |  |      |  |         |  |     | github.com/crossplane/crossplane/apis/apiextensions/v1beta1.StringTransform
|      |        |  |      |  |         |  |     |
|      |        |  |      |  |         |  |     |
|      |        |  |      |  |         |  |     |
source.PatchSets[].Patches[].Transforms[].String*.???
target.PatchSets[].Patches[].Transforms[].String*.Break
|      |        |  |      |  |         |  |     | |
|      |        |  |      |  |         |  |     | | *github.com/crossplane/crossplane/apis/apiextensions/v1.StringTransformBreak
|      |        |  |      |  |         |  |     |
|      |        |  |      |  |         |  |     | github.com/crossplane/crossplane/apis/apiextensions/v1.StringTransform
|      |        |  |      |  |         |  |
|      |        |  |      |  |         |  | *github.com/crossplane/crossplane/apis/apiextensions/v1.StringTransform
|      |        |  |      |  |         |
|      |        |  |      |  |         | github.com/crossplane/crossplane/apis/apiextensions/v1.Transform
|      |        |  |      |  |
|      |        |  |      |  | []github.com/crossplane/crossplane/apis/apiextensions/v1.Transform
|      |        |  |      |
|      |        |  |      | github.com/crossplane/crossplane/apis/apiextensions/v1.Patch
|      |        |  |
|      |        |  | []github.com/crossplane/crossplane/apis/apiextensions/v1.Patch
|      |        |
|      |        | github.com/crossplane/crossplane/apis/apiextensions/v1.PatchSet
|      |
|      | []github.com/crossplane/crossplane/apis/apiextensions/v1.PatchSet
|
| github.com/crossplane/crossplane/apis/apiextensions/v1.CompositionSpec

Cannot match the target field with the source entry: "Break" does not exist.
exit status 1
apis/generate.go:44: running "go": exit status 1
21:14:16 [FAIL]
make[2]: *** [build/makelib/golang.mk:240: go.generate] Error 1
make[1]: *** [build/makelib/common.mk:434: generate] Error 2
make: *** [build/makelib/common.mk:439: reviewable] Error 2
```

**Comment by negz** [comment]
I've marked this ready for review after manual testing (per PR description). Note that I think the failing code coverage test is misleading. It shows overall project coverage dropping by ~1.1% in this PR - that's probably true but it's largely because human-written code that was covered by unit tests has now become auto-generated code that is not.

File: hack/duplicate_api_type.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
```
Comment (negz) [inline hack/duplicate_api_type.sh:1]: I've run this on both Linux and MacOS and it appears to work fine.

File: apis/pkg/meta/v1alpha1/zz_generated.configuration_types.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
+/*
+Copyright 2020 The Crossplane Authors.
```
Comment (ezgidemirel) [inline apis/pkg/meta/v1alpha1/zz_generated.configuration_types.go:2]:
```suggestion
Copyright 2022 The Crossplane Authors.
```

File: apis/pkg/meta/v1alpha1/zz_generated.provider_types.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2020 The Crossplane Authors.
```
Comment (ezgidemirel) [inline apis/pkg/meta/v1alpha1/zz_generated.provider_types.go:2]:
```suggestion
Copyright 2022 The Crossplane Authors.
```

File: internal/controller/apiextensions/composition/revision.go [hunk 4]
```diff
@@ -55,193 +55,8 @@ func NewCompositionRevision(c *v1.Composition, revision int64, compSpecHash stri
...
-		rs.PatchSets[i] = NewCompositionRevisionPatchSet(cs.PatchSets[i])
-	}
-
-	for i := range cs.Resources {
-		rs.Resources[i] = NewCompositionRevisionComposedTemplate(cs.Resources[i])
-	}
-
+	conv := v1.GeneratedRevisionSpecConverter{}
+	rs := conv.ToRevisionSpec(cs)
+	rs.Revision = revision
```
Comment (ezgidemirel) [inline internal/controller/apiextensions/composition/revision.go:60]: I loved this simplified and safer version :)

**Review by ezgidemirel — APPROVED** [review]

File: apis/pkg/meta/v1alpha1/zz_generated.configuration_types.go (hunk 2 above)
Comment (negz) [reply apis/pkg/meta/v1alpha1/zz_generated.configuration_types.go:2]: I'm going to leave these as-is.

While the new files were obviously added this year, they're "generated" copies of files added in 2020. To have them include the correct year in the copyright we'd either have to teach the (very basic) shell script that makes the copies to alter the dates, or update the copyright on the original files.

**Comment by jbw976** [comment]
This was a nice automated solution to a problem that has been plaguing the code base for awhile @negz, nice work!! 💪

**Pull request merged**